File: backend/deploy.rb
#!/usr/bin/env ruby

# Run from the local directory.

d3 = File.read("../d3.v4.min.js")
d3_selection = File.read("../d3-selection-multi.v0.4.min.js")
uuid = `uglifyjs -c -m sort,toplevel -- ../uuid.js`
pressing_game = `uglifyjs -c -m sort,toplevel -- ../pressing-game.js`

`cleancss -o ../../release/pressing-game.css ../pressing-game.css`

css = File.read("../../release/pressing-game.css");
index = File.read("../index.htm")

script = "<style>#{css}</style>"

[d3, d3_selection, uuid, pressing_game].each do |file|
   # use separate script tags for each file
  script += "<script type=\"text/javascript\">#{file}</script>"
end

scr = index.gsub!(/<!--[\s]+replace-start[\s\S]*replace-end[\s]+-->/, script)

f = File.open('../../release/index-inlined-scripts.htm', 'w+')
f.write(scr)
f.close

htm = `html-minifier --remove-comments --collapse-whitespace --case-sensitive ../../release/index-inlined-scripts.htm > ../../release/index-min.htm`

favicon = `cp ../favicon.ico ../../release/favicon.ico`

gzip = `gzip -9 -c ../../release/index-min.htm > ../../release/index.htm`

# TO DO:
# Rewrite in Javascript, since we are using node.js packages anyway.
# List deployment script dependencies somewhere in the repo.
# Upload with to S3 with public permissions and Content-Encoding=gzip.
# Add W3C validation API (html/css) and lint validation for js (uglify *may* include lint).

File: dev/backend/deploy.js
var fs = require('fs'),
	uglifyJS = require('uglify-js'),
	minify = require('html-minifier').minify,
	cleanCSS = require('clean-css'),
	util = require('util'),
	exec = require('child_process').exec;//,
	//Regex = require('regex');

var d3 = fs.readFileSync('d3.v4.min.js', 'utf-8'),
	d3_selection = fs.readFileSync('d3-selection-multi.v0.4.min.js', 'utf-8'),
	uuid = uglifyJS.minify('uuid.js', {mangle: {toplevel:true} }).code,
	pressing_game = uglifyJS.minify('pressing-game.js', {mangle: {toplevel:true} }).code,
	css = new cleanCSS().minify(fs.readFileSync('pressing-game.css', 'utf-8')).styles;

var js =	"<style>" + css + "</style>" +
			//"<script type= text/javascript>" + d3 + "</script>" +
			//"<script type= text/javascript>" + d3_selection + "</script>" +
			//"<script type= text/javascript>" + uuid + "</script>" +
			"<script type= text/javascript>"+ pressing_game + "</script>";

var indexInlined = fs.readFileSync('index.htm', 'utf-8');

var indexMinified = minify(indexInlined, {
		caseSensitive : true,
		collapseWhiteSpace : true
	});

var R = new RegExp('<!--\\s+replace-start\\s+-->(.*)<!--\\s+replace-end\\s+-->'); // /<!--[\s]+replace-start[\s\S]*replace-end[\s]+-->/);

var src = indexMinified.replace('<!--\\s+replace-start\\s+-->(.*)<!--\\s+replace-end\\s+-->', js);
console.log(src);


/*
fs.writeFileSync('../release/index-min.htm', src);

var favicon = fs.readFile('favicon.ico', 'utf-8');
fs.writeFileSync('../release/favicon.ico', favicon);

exec('gzip -9 -c ../release/index-min.htm > ../release/index.htm', function (error, stdout, stderr) {

  if (error !== null) {
    console.log('exec error: ' + error);
  }

});*/
